Implement unified interface to install scripts

kxtcd950 opened the pull request:
I've implemented a unified interface to the install scripts for the components - "hassbian-config"
This takes the form of a single script which uses the (slightly modified) install scripts previously used to be able to install and show information about the packages (suites, in new "hassbian-config" terms).
The script itself is nothing terribly special, it just uses the installers to be able to show which suites are available, information about those suites, and to install them (dynamically retrieved so that this script can remain unchanged despite packages being added).
Note that no attempt at an uninstall command for hassbian-config has been attempted due to there being no uninstallers present for the installers as historically presented.

It's very simplistic, and if accepted, will require a change in the install systemd unit file to reference the hassbian-config script rather than calling install_homeassistant.sh as the unit file does now.

The changes to the install_.sh files have been kept to a minimum, and no changes have been made in the code which actually performs the installs - only to wrap it with a bash function declaration which the hassbian-config script itself calls after sourcing the install_.sh file as required.

The requirements for install_*.sh scripts are that:

* the filename is of a normalised pattern - install_*.sh - for discovery purposes.
* four functions exist, named after what would be captured with the regexp install_(.*).sh when presented with the filename. This would yield a captured text of "homeassistant" for an suite file named install_homeassistant.sh.

The four functions are ($1 means the text as captured from the regexp against the filename):

$1-show-short-info (shown in the list of suites available with the "show" command)
$1-show-long-info (shown in the information about a single suite from the "info" command)
$1-show-copyright-info (shown in the info when a single suite is listed with the "show" command; also shown in the installation process)
$1-install-package (the function which installs the package, as the script used to when run standalone in response to the "install" command)

These are called as required by the "hassbian-config" script as indicated above.

Sorry also about the messed up commit history; I'm not really that familiar with git (but I'm learning fast!)

kxtcd950 committed:
Initial commit for hassbian-config script.
The hassbian-config script is a framework for configuring the installed packages on the system.
It utilises the install scripts with minor modifications for inclusion into the framework.

hassbian-config defines three operations: install, show and info

install will install the named software package (hassbian-config calls this a "suite")
show will show all available installers.
info will show information about the suite selected.

As of this version, there is no kind of state kept by hassbian-config; this will likely need to be changed.

kxtcd950 committed:
Add documentation for hassbian-config.

kxtcd950 committed:
Added comment as to why the homeassistant suite is not in the list of available suites.

kxtcd950 committed:
Attempt to commit all local changes (I'm not familiar with git, sorry).

kxtcd950 committed:
Added stanzas to all suite installers; updated install_homeassistant.sh to be "just" another suite.

homeassistant commented:
Hi @kxtcd950, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;hassbian-scripts#11).

Once you do that we will be able to review and accept this pull request.

Thanks!

Landrash commented:
This is awesome! Going to do some testing and hopefully merge this is as soon as possible

kxtcd950 committed:
Untab file - somehow tabs had crept in, not spaces.

Landrash commented:
@kxtcd950 I love this and I'll merge it in this weekend.  I'm going to make a dev branch and and try to follow the same plan as we do with the HA release. Every other week for larger changes such as this and patches in between.

Landrash commented on install_homeassistant.sh:
```diff
@@ -1,8 +1,19 @@
 #!/bin/sh
```
There shouldn't need to be any change to this script since it's the one ran when we create the image. It's not run after the image is created so the tool shouldn't need to care about this script. Might move this script into the files folder but no change you need to do.

kxtcd950 replied:
I was thinking about this too - it's a breaking change for first installation, which I'm not happy with.  I'm sorta thinking that having it available isn't a bad idea, as it allows for reinstalls of homeassistant to happen from the hassbian-config script; otherwise users have to find out how to perform this one install themselves.  I did inhibit it from being viewed as a "suite" though, because it shouldn't ordinarily be run...
With a little more effort, it may be possible to detect that install_homeassistant.sh is being run at first-run time, and make it work as it has previously.  It's a case of changing the very last line of the script.
If you're interested, I could look into making this change; but I'm at work and so it won't be a speedily forthcoming one :(

kxtcd950 replied:
Ok, it's not as difficult as I first thought.  I've just pushed a new change in so that the install_homeassistant.sh script when run standalone acts as if it were the old installer.  Bonus: it still fits in the hassbian-config mechanism so it can be used from there too.
This also fixes the install_homeassistant.sh script's hashbang which somehow hadn't been edited to use bash, but /bin/sh - which precluded the whole lot working.

Landrash reviewed (changes requested):
One small comment.
More might come after I've tested it properly.

kxtcd950 committed:
Fixup tabs
Correct /bin/sh stanza which precludes the "function" keyword
Restore old functionality if installer is run standalone.

Landrash commented on install_homeassistant.sh:
```diff
@@ -1,10 +1,10 @@
 
 function homeassistant-show-long-info {
-	echo "Installs the libcec package for controlling CEC devices from this Pi"
+    echo "Installs the libcec package for controlling CEC devices from this Pi"
```
Not the libcec package.

Landrash reviewed (approved):
All these changes are awesome!
Love this and hope to have the time to give it a proper run down as soon as possible. (This weekend)

kxtcd950 committed:
Change package long info text to reflect reality.

balloob commented:
W00t 🍻  love it

balloob commented:
This is some advanced shell scripting 👍  This PR fixes #9

I also created a follow up feature request from me: https://github.com/home-assistant/hassbian-scripts/issues/12 . Would love some input to see if shell scripts will be flexible enough to implement such features in the future.

[Radiodan's config tool](https://github.com/radiodan/provision) has a folder per task, which is nice because that way if we depend on files, we can have it all nicely organized together.

kxtcd950 committed:
Remove local testing variable.

balloob commented:
@Landrash does the current hassbian image pull this repo? If so, would merging this to master break all new image installs ?

Landrash commented:
Currently the image pulls this repo yes. I'm planning on merging it into a
new branch instead.

I'll have time for it in a few hours.

On Wed, Feb 8, 2017, 18:26 Paulus Schoutsen <notifications@github.com>
wrote:

> @Landrash <https://github.com/Landrash> does the current hassbian image
> pull this repo? If so, would merging this to master break all new image
> installs ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/hassbian-scripts/pull/11#issuecomment-278398687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABrPCnOBx7GaNTDiGqCnFrEvmKEgcBcKks5rafqngaJpZM4L4Ke6>
> .
>
